dev/DynamicDependency/MddDetourPackageGraph.cpp

```diff
@@ -0,0 +1,104 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+#include "pch.h"
+
+#include "MddDetourPackageGraph.h"
+
+#include "PackageGraphManager.h"
+
+#include <../Detours/detours.h>
+
+// Windows provides HRESULT_FROM_WIN32() but not the reverse. We need that for compat reasons.
+// Define WIN32_FROM_HRESULT() to covert an HRESULT to a Win32 Error Code if the HRESULT's facility
+// is FACILITY_WIN32. See https://devblogs.microsoft.com/oldnewthing/20061103-07/?p=29133 for more details
+#if !defined(WIN32_FROM_HRESULT)
+inline DWORD Win32FromHResult(const HRESULT hr)
+{
+    if (HRESULT_FACILITY(hr) == FACILITY_WIN32)
+    {
+        return static_cast<DWORD>(HRESULT_CODE(hr));
+    }
+    else
+    {
+        return static_cast<DWORD>(hr);
+    }
+}
+#define WIN32_FROM_HRESULT(hr) Win32FromHResult(hr)
+#endif
+
+static LONG (WINAPI * TrueGetCurrentPackageInfo)(
+    const UINT32 flags,
+    UINT32* bufferLength,
+    BYTE* buffer,
+    UINT32* count) = GetCurrentPackageInfo;
+
+LONG WINAPI DynamicGetCurrentPackageInfo(
+    const UINT32 flags,
+    UINT32* bufferLength,
+    BYTE* buffer,
+    UINT32* count);
+
+static LONG (WINAPI * TrueGetCurrentPackageInfo2)(
+    const UINT32 flags,
+    PackagePathType packagePathType,
+    UINT32* bufferLength,
+    BYTE* buffer,
+    UINT32* count) = GetCurrentPackageInfo2;
+
+LONG WINAPI DynamicGetCurrentPackageInfo2(
+    const UINT32 flags,
+    PackagePathType packagePathType,
+    UINT32* bufferLength,
+    BYTE* buffer,
+    UINT32* count);
+
+HRESULT WINAPI MddDetourPackageGraphInitialize() noexcept
+{
+    // Detour package graph APIs to our implementation
+    FAIL_FAST_IF_WIN32_ERROR(DetourUpdateThread(GetCurrentThread()));
```

Comment:
`DetourUpdateThread(GetCurrentThread())` is a no-op, see https://github.com/microsoft/Detours/issues/151 and consider using `CreateToolhelp32Snapshot` to enumerate and update all other threads.

Reply:
> `DetourUpdateThread(GetCurrentThread())` is a no-op
Ouch! I see what you mean

```
LONG WINAPI DetourUpdateThread(_In_ HANDLE hThread)
{
    LONG error;

    // If any of the pending operations failed, then we don't need to do this.
    if (s_nPendingError != NO_ERROR) {
        return s_nPendingError;
    }

    // Silently (and safely) drop any attempt to suspend our own thread.
    if (hThread == GetCurrentThread()) {
        return NO_ERROR;
    }
```

That `if (hThread == GetCurrentThread())` is even more interesting when you realize [GetCurrentThread() returns a pseudo handle](https://docs.microsoft.com/windows/win32/api/processthreadsapi/nf-processthreadsapi-getcurrentthread#return-value). It wouldn't match if you passed the actual handle to the current thread.

I got this from Detours' wiki and samples. [UndockedRegFreeWinRT](https://github.com/microsoft/xlang/tree/master/src/UndockedRegFreeWinRT/UndockedRegFreeWinRT) has the same problem.

This will need more noodling. I'll open an issue for follow up in a separate PR.

I'll also follow up with @kythant  and @BenJKuhn re UndockedRegFreeWinRT.

Thanks for the feedback!

Reply:
If you pass an actual thread handle, then the process deadlocks. I've hit that case myself and opened a PR for it, but due to Detours targeting XP, the PR can't be merged (`GetThreadId` got added in Vista).

The issue I linked also mentions that all the samples are incorrect, so I wouldn't be surprised if most usages out there hit the same trap.

Reply:
I filed [DetourUpdateThread(GetCurrentThread()) is a no-op #723](https://github.com/microsoft/xlang/issues/723) in the xlang project for UndockedRegFreeWinRT

Reply:
[MSIX Dynamic Dependencies PR follow ups #403](https://github.com/microsoft/ProjectReunion/issues/403)